### PR TITLE
ECS Refactor.

### PR DESCRIPTION
**Pull request by Richardn2002**
*Putting my plans and designs here later.*

### CODE REVIEW

File: inox2d/src/node/components.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+pub(crate) mod zsort;
+
+pub use composite::Composite;
+pub use drawable::Drawable;
+pub use simple_physics::SimplePhysics;
+pub use textured_mesh::TexturedMesh;
+
+pub(crate) use deform_stack::DeformStack;
+pub(crate) use transform_store::TransformStore;
+pub(crate) use zsort::ZSort;
```
Comment (Speykious) [inline inox2d/src/node/components.rs]: Given the respective sizes of these modules, I think all of the structures defined in there should live in this single file instead. Right now it feels artificially sparse, the individual structures don't have any complex logic on their own that would be contained in separate files, making the modules shallow.

Comment (Richardn2002) [reply inox2d/src/node/components.rs]: This is a valid point, as component definition files indeed should only contain data and no logic.
And I should probably put `DeformStack` logic to somewhere else, likely under `render`, cause this is indeed *some* logic.

File: inox2d/src/node/components/deform_stack.rs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+use glam::Vec2;
+
+use crate::math::deform::{linear_combine, Deform};
+use crate::node::InoxNodeUuid;
+use crate::params::ParamUuid;
+use crate::puppet::{InoxNodeTree, World};
+
+/// Source of a deform.
+#[derive(Hash, PartialEq, Eq, Copy, Clone)]
+pub(crate) enum DeformSrc {
```
Comment (Speykious) [inline inox2d/src/node/components/deform_stack.rs]: I think I'd prefer `DeformSource` as a name

File: examples/render-opengl/src/main.rs [hunk 3]
```diff
@@ -41,12 +41,18 @@ fn main() -> Result<(), Box<dyn Error>> {
...
 	tracing::info!(
 		"Successfully parsed puppet: {}",
 		(model.puppet.meta.name.as_deref()).unwrap_or("<no puppet name specified in file>")
 	);
 
+	tracing::info!("Setting up puppet for transforms, params and rendering.");
+	model.puppet.init_transforms();
+	model.puppet.init_rendering();
+	model.puppet.init_params();
+	model.puppet.init_physics();
```
Comment (Speykious) [inline examples/render-opengl/src/main.rs:54]: So if this is all called at the same time, why are they separate functions in the first place?

File: inox2d/src/node/drawables.rs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	/// - Panicks if component combinations invalid.
+	pub(crate) fn new(id: InoxNodeUuid, comps: &'comps World) -> Option<Self> {
+		let drawable = match comps.get::<Drawable>(id) {
+			Some(drawable) => drawable,
+			None => return None,
+		};
+		let transform = &comps
+			.get::<TransformStore>(id)
+			.expect("A drawble must have an associated transform.")
+			.absolute;
```
Comment (Speykious) [inline inox2d/src/node/drawables.rs:49]: It's weird that `.get().expect()` is needed for `TransformStore`. Doesn't literally every node need to store transform data? Ideally this function should just get it from the `InoxNode` itself or something similar.

File: inox2d/src/node/drawables.rs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+		let textured_mesh = comps.get::<TexturedMesh>(id);
+		let composite = comps.get::<Composite>(id);
+
+		match (textured_mesh.is_some(), composite.is_some()) {
+			(true, true) => panic!("The drawable has both TexturedMesh and Composite."),
+			(false, false) => panic!("The drawable has neither TexturedMesh nor Composite."),
```
Comment (Speykious) [inline inox2d/src/node/drawables.rs]: A node having both a `TexturedMesh` and a `Composite` is something that is up to the model, so it's something that can actually happen. Crashing doesn't make sense here.
I propose that we return something that seems wrong instead of panicking. For example `(false, true) | (true, true) => Some(DrawableKind::Composite(...))` (which gives priority to the composite and ignores the textured mesh) and `(false, false) => None`.
Then somewhere else during initialization, we can log an error with tracing to warn the user of a malformed of corrupted model, and even go as far as to say that the textured mesh will be ignored.

File: inox2d/src/node/drawables.rs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+	pub transform: &'comps Mat4,
+	pub drawable: &'comps Drawable,
+	pub data: &'comps Composite,
+}
+
+impl<'comps> DrawableKind<'comps> {
+	/// Tries to construct a renderable node data pack from the World of components:
+	/// - `None` if node not renderable.
+	/// - Panicks if component combinations invalid.
+	pub(crate) fn new(id: InoxNodeUuid, comps: &'comps World) -> Option<Self> {
```
Comment (Speykious) [inline inox2d/src/node/drawables.rs]: I'm not sure if this is the best place for this function or if it's even the right approach. It doesn't feel right to have to look at the components of a node only for it to return an enum that we have to then match against instead of doing it more directly. I don't have a nicer solution right now though but I'm probably gonna think about refactoring this part in the future.

File: inox2d/src/node/components/drawable.rs [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+	pub const VALUES: [BlendMode; 7] = [
+		BlendMode::Normal,
+		BlendMode::Multiply,
+		BlendMode::ColorDodge,
+		BlendMode::LinearDodge,
+		BlendMode::Screen,
+		BlendMode::ClipToLower,
+		BlendMode::SliceFromLower,
+	];
+}
```
Comment (Speykious) [inline inox2d/src/node/components/drawable.rs]: Why is this here? It's not even used :v
(my guess is you intended to use it for something but forgot to remove it)

File: inox2d/src/render.rs [hunk 8]
```diff
@@ -1,378 +1,386 @@
...
 
-		for uuid in nodes.zsorted_root() {
-			let node = nodes.get_node(uuid).unwrap();
+	/// Reset all `DeformStack`.
+	pub(crate) fn reset(&mut self, nodes: &InoxNodeTree, comps: &mut World) {
+		for node in nodes.iter() {
+			if let Some(DrawableKind::TexturedMesh(..)) = DrawableKind::new(node.uuid, comps) {
+				let deform_stack = comps
+					.get_mut::<DeformStack>(node.uuid)
+					.expect("A TexturedMesh must have an associated DeformStack.");
```
Comment (Speykious) [inline inox2d/src/render.rs]: I see a few of these `DeformStack` component fetches. The best thing to do here in case it doesn't exist is to do nothing. This is again a case where you'd log an error at initialization time if the puppet is wrong.

File: inox2d/src/render.rs [hunk 9]
```diff
@@ -1,378 +1,386 @@
...
+		VertexBuffers::vec_vec2_as_vec_f32(&self.vertex_buffers.uvs)
+	}
+	/// Memory layout: `[[i0, i1, i2], [i0, i1, i2], ...]`
+	pub fn get_raw_indices(&self) -> &[u16] {
+		self.vertex_buffers.indices.as_slice()
+	}
+	/// Memory layout: `[[dx, dy], [dx, dy], ...]`
+	pub fn get_raw_deforms(&self) -> &[f32] {
+		VertexBuffers::vec_vec2_as_vec_f32(&self.vertex_buffers.deforms)
 	}
```
Comment (Speykious) [inline inox2d/src/render.rs]: These getters aren't necessary. They add noise more than anything right now. Better just use the code that's inside them on the one call site where they're used and keep every field of the `RenderCtx` as public.

File: inox2d/src/render.rs [hunk 10]
```diff
@@ -1,378 +1,386 @@
...
-	/// The render pass.
-	///
-	/// Logical error if this puppet is not from the latest prepared model.
-	fn render(&self, puppet: &Puppet);
-	/// Finish one render pass.
-	fn on_end_scene(&self);
-	/// Actually make results visible, e.g. on a screen/texture.
-	fn draw_scene(&self);
+	/// Ref impl: Upload textures.
+	fn new(params: Self::NewParams, model: &Model) -> Result<Self, Self::Error>;
```
Comment (Speykious) [inline inox2d/src/render.rs]: This function is not used polymorphically inside the library, so I think it's better to not specify any need for it. It'll remove the need for the `NewParams` associated type and make the `new` function of every renderer more flexible.

File: inox2d/src/render.rs [hunk 11]
```diff
@@ -1,378 +1,386 @@
...
+/// This does not guarantee the display of a puppet on screen due to these possible reasons:
+/// - Only provided `InoxRenderer` method implementations are called.
+/// For example, maybe the caller still need to transfer content from a texture buffer to the screen surface buffer.
+/// - The provided `InoxRender` implementation is wrong.
+/// - `puppet` here does not belong to the `model` this `renderer` is initialized with. This will likely result in panics for non-existent node uuids.
+pub fn draw<T: InoxRenderer>(renderer: &T, puppet: &Puppet) {
+	renderer.on_begin_draw(puppet);
+	renderer.draw(puppet);
+	renderer.on_end_draw(puppet);
+}
```
Comment (Speykious) [inline inox2d/src/render.rs]: I don't think it's a good idea to define this function. Calling `on_begin_draw` then `draw` then `on_end_draw` is more easy to understand than calling a separate `inox2d::render::draw(...)` function which itself has to be inspected.

File: inox2d-opengl/src/gl_buffer.rs [hunk 12]
```diff
@@ -11,71 +9,50 @@ unsafe fn upload_array_to_gl<T>(gl: &glow::Context, array: &[T], target: u32, us
...
+/// Create a vertex array. Initialize vertex, uv, deform and index buffers, upload content and attach them to the vertex array. Return the array.
+///
+/// # Errors
+///
+/// This function will return an error if it couldn't create a vertex array.
+///
+/// # Safety
+///
+/// No prerequisites.
+pub unsafe fn setup_gl_buffers(
```
Comment (Speykious) [inline inox2d-opengl/src/gl_buffer.rs]: If there are no prerequisites, I guess the function shouldn't be `unsafe` and it should just have an `unsafe` block inside.

**Review by Speykious — CHANGES_REQUESTED** [review]
I think overall I like this change. There's a few things here and there that I commented on but otherwise it's nice. I'll want to do another refactor pass on this when it's finalized and merged to improve on the doc comments you provided, and then later see what I can do about this very hacky `AnyVec` implementation. I'm probably going to replace it with an actual page-allocated arena, if all our supported targets allow for it.

File: examples/render-opengl/src/main.rs (hunk 3 above)
Comment (Richardn2002) [reply examples/render-opengl/src/main.rs:54]: Related to the comment below: It is actually possible to:
- Not call any of these init functions, only load whatever is in the inp/inx file into RAM. Potential applications include a simple parser.
- Only init transform + render, so displaying a static model is possible. Potential applications include a model previewer.
- Only init transform + render + param, to render a model without physics. Potential applications include a renderer with low perf settings.
- Init all and enjoy all.

Such modularized init is a nice byproduct of the refactor so I enable it.

File: inox2d/src/node/drawables.rs (hunk 4 above)
Comment (Richardn2002) [reply inox2d/src/node/drawables.rs:49]: See [above](https://github.com/Inochi2D/inox2d/pull/86#discussion_r1698746606).

File: inox2d/src/node/components/drawable.rs (hunk 7 above)
Comment (Richardn2002) [reply inox2d/src/node/components/drawable.rs]: Fun fact: this is used by the WGPU renderer for some reason. Tho the WGPU renderer is seriously out of date now...

Comment (Speykious) [reply inox2d/src/node/components/drawable.rs]: Let's keep it for now until we revive the WGPU renderer then. I'll probably move that to inox2d-wgpu eventually

File: inox2d/src/node/drawables.rs (hunk 5 above)
Comment (Richardn2002) [reply inox2d/src/node/drawables.rs]: A node having both components is out of Inochi2D spec. It is us who choose to represent `class Part : Drawable` and `class Composite : Drawable` in the D ref impl with composition in Rust. If it is type wise not possible to describe a node with both (you can't override a same function twice!) in D, I guess it is not proper for Inox2D core to do so.

If you are persuaded that this is out of spec, and still wants to keep such behavior as Inox2D extension, then my opinion on such an extension is: Keep having both of these components a signal of a bug on our end or an improper usage from the end user. Use components of other types for representing such special nodes that act as both textured mesh and composite.

File: inox2d/src/node/drawables.rs (hunk 6 above)
Comment (Richardn2002) [reply inox2d/src/node/drawables.rs]: Related to [above](https://github.com/Inochi2D/inox2d/pull/86#discussion_r1698768218). This "component combination to enum variants" mapping is for following the Inochi2D spec. There should be a clear way of detecting "official" nodes, and then executing them in the "official" way.

Instead of removing this, what should be done further is a way to inject a end-user-provided routine to handle custom nodes, and before we hand over a node to a custom routine, we still need a way to extract the non-custom part and do what inox2d is supposed to do, which is what we have here, this `new()`.

File: inox2d/src/render.rs (hunk 8 above)
Comment (Richardn2002) [reply inox2d/src/render.rs]: My initial response would be: If rendering is initialized, then a `TexturedMesh` MUST have a `DeformStack`, because without a deform (outputted by the respective stack), the shader just don't work. This is not a sign of wrong puppet. This is a sign of wrong rendering initialization logic, which is our problem. So probably better to panic then to continue silently.

My second response: Oh, you give me the insight that we can NOT create a `DeformStack` for a `TexturedMesh` that is not deformed by any parameter (or, the parameter system is not initialized at all), and leave a default buffer region with zeroes for the GPU. In this case yeah we can just skip `None`.

File: inox2d/src/render.rs (hunk 9 above)
Comment (Richardn2002) [reply inox2d/src/render.rs]: Now I think it, such interfaces are loosely inspired by the "slab allocator" talk by Luna, where node data for GPU can be efficiently re-ordered for efficient batched rendering. In such case, actual data layout will be complex, but renderer implementers still want to see a contiguous memory region for upload. And this is where such "adapter" functions come in.

But true, we are so far away from actually having an idea of how to design such an allocator, so probably a good idea to remove these getters for now. We will have a better idea of what the interface should be when internal memory layouts really begin to change.

File: inox2d/src/render.rs (hunk 11 above)
Comment (Richardn2002) [reply inox2d/src/render.rs]: This function is there for restricting the visibility of crate internals. I don't want the trait containing the auto implemented `.draw()` (currently under `InoxRendererCommon`) to be public for end users to `use` and get the `.draw()`. This gives a wrong idea of `InoxRendererCommon` can be implemented themselves.

As for the urge of inspecting everything, my response to suppressing that urge is to use intensive comments (something I had already done), so that users can start to believe as long as they implement what are prompted to implement, inox2d will do the rest, and do what they expect.

I will still admit that I am not an expert in all Rust features related to `trait`s, so feel free to offer alternatives controlling the visibility, or explain why visibility is not important here.

**Comment by Richardn2002** [comment]
What does "another refactor pass" mean :eyes:? IMO we really should be doing MeshGroups and other rendering backends and even game engine integrations afterwards.
Sure the `World` implementation can be improved. I would like more efficient iteration, and, if possible, dark magic allowing obtaining two mutable handles to two components of different types, or even just two different components. But anyways I don't think reimplementing `World` would break too much stuff and cause a refactor.
The idea of "systems" and the process of systems applying to node data in order can be formalized, yes. If you want to call this a "refactor" I am also fine with it.

Btw, my `AnyVec` is hacky, intentionally, because I want to steal all the work done by Rust `std` team for `Vec`. We all have seen some perf results and we all know matching `std` performance is not easy :eyes:. And it is not that `Vec` can't use a custom allocator. I do suggest giving more thoughts to the idea of reusing `Vec` as I am doing now.

File: inox2d/src/node/drawables.rs (hunk 5 above)
Comment (Speykious) [reply inox2d/src/node/drawables.rs]: It being out of the spec got past me, but it is not really a good reason to add points of panic. inox2d is only responsible for moving puppets, it's weird and unexpected for such a library to crash the end user's program. In our case it's better to continue with a wrong-ish state because the end user can still run their application without it crashing, which would be more annoying. Think textures not loading being replaced by black-magenta default textures instead of blowing up and not running. So I maintain that if it has a problem it should ideally log it and move on as best it can.

File: inox2d/src/render.rs (hunk 11 above)
Comment (Speykious) [reply inox2d/src/render.rs]: I think what we can do is to rename `InoxRendererCommon` to `InoxRendererExt` (`Ext` is a common suffix for extension crates), making it clear that it's an extension trait. That way, it being public will be no problem as it'll be clear that it's automatically added functionality for any renderer, and that you can't (and are not supposed to) implement any of that yourself.

**Comment by Speykious** [comment]
Looks good to me. Now we just need to revive inox-wgpu

**Comment by Speykious** [comment]
After discussing it on Discord, we decided it was best to get rid of the WGPU renderer for now. Neither of us can maintain it because we don't know WGPU at all, and once either of us learns it, we're gonna need to rewrite it from scratch anyway because of this refactor and possble future ones to come.

**Review by Speykious — APPROVED** [review]
We're good now
